### PR TITLE
Fixed an issue with quickInfo crashing on a function property returned from generic function

### PR DESCRIPTION
**Pull request by Andarist**
Fixes https://github.com/microsoft/TypeScript/issues/49048

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -4020,7 +4020,10 @@ namespace ts {
                         }
                     });
                 }) : undefined;
-                let res = firstVariableMatch ? [firstVariableMatch, ...additionalContainers, container] : [...additionalContainers, container];
+                let res = firstVariableMatch ? [firstVariableMatch, ...additionalContainers] : [...additionalContainers];
+                if (!container.valueDeclaration || container.valueDeclaration.kind !== SyntaxKind.FunctionDeclaration || !isExpandoFunctionDeclaration(container.valueDeclaration)) {
```
Comment (Andarist) [inline src/compiler/checker.ts:4024]: I'm not super comfortable with this line of code. My goal was to **preserve** the very same info text that we get when requesting this based on the generated `.d.ts`. We can check this situation [here](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220510#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXzDigxAAUYcAHAZwB4AVAPgAoMAueegSg4G8AoeEPjMe8AG44swANyDhhEMRAA1WFiioM1DqPgBeRhKmz+AXzn8A9FfgALHOJAwsqAObxHz+AANFytRdNbR94YBwQalQAcgx4AHccGABre2cQfn8ScipqZl4zLgA6LNV1YLyuOSA).

As we might notice this **doesn't** include `getProps` as that is being elided in the `.d.ts`.

Maybe I shouldn't try accomplishing this at all, and I should be fine with this "inconsistency", and thus I should just be fine with returning `(property) getProps.createVariants: () => void` here.

IIRC that can be achieved by just handling `SyntaxKind.FunctionDeclaration` [here](https://github.dev/microsoft/TypeScript/blob/9236e39374c0ec9a1e3f9894af4fb9eb34ba0021/src/compiler/checker.ts#L10064). However, I feel that in such a case the function name's probably should be adjusted - if you prefer that solution, let me know and if you could suggest how this should be renamed I would appreciate it (if that's a suitable solution in the first place).

Either way... in an attempt to be consistent about the content of this text between those scenarios I've followed the logic of the `.d.ts` generation and I've found out about the `isExpandoFunctionDeclaration` check being involved. I'm not sure what it exactly means though and including it here changes some existing baselines which isn't good.

At first I've thought that maybe they should actually be adjusted if I'm trying to be "consistent" here but I've analyzed them, checked their `.d.ts` and stuff and I **think** that they shouldn't be changed here (although I'm not completely sure about that, I wasn't exactly sure how properly to validate this for all of those adjusted baselines). So I feel like maybe this check here isn't exactly correct but I'm unsure how to adjust it right now.

Btw. any kind of a condition added to fix this should also cover function expressions case - I probably should add a similar test case for this scenario as it doesn't work either ([here](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220510#code/MYewdgzgLgBMBOBTAhlRAFeIAOEYF4YAeAFQBoA+ACigC4YSBKAimAbwCgY5xoYBzRFEw48hKs3ys2AXy49IsBCjQA1ZPACWyMFDEwJLdnPmDhWXADplqROq069BOElv3tuiAG55SKAFd4MAEhEVwfEwB6SJgACxAAN0QtMH4YROSYAAMbNQ0PPSyYABMQRAgwAHJYAHcQeABrOOTEDlyMCwgqWUZrVzyHTwkfIA)).

Comment (DanielRosenwasser) [reply src/compiler/checker.ts:4024]: I have no idea about if the fix is right. The thing I'm more surprised about is the fact that any of this fires in a `.ts` file. IIRC, we're only supposed to support "namespace"-y function declarations at the top level.

Comment (Andarist) [reply src/compiler/checker.ts:4024]: What kind of support do u mean? I've rechecked all TS versions available in the playground and the pattern seems to be typechecked correctly in all of them (the quick info crashes in all of them though). 🤔 From what I understand this is required to make stuff like `.defaultProps` to work in React.

Comment (Andarist) [reply src/compiler/checker.ts:4024]: > An expando function property actually equivalent to

Out of curiosity - is "expando function property" specifically about function+namespace combination? Or Special Property Assignment on functions always treated/named as expando too?

> I think the getProps. should be retained; I don't think it matters that much whether it reads function or (property)

My original hesitation before including the namespace was that in the generated `.d.ts` there won't be any namespace here, it's being completely elided:
```ts
// .ts
function createProps<T>(t: T) {
  function getProps() {}
  function createVariants() {}

  getProps.createVariants = createVariants;
  return getProps;
}

// .d.ts
declare function createProps<T>(t: T): {
    (): void;
    createVariants: () => void;
};
```

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -10063,7 +10063,19 @@ namespace ts {
 
         // The outer type parameters are those defined by enclosing generic classes, methods, or functions.
         function getOuterTypeParametersOfClassOrInterface(symbol: Symbol): TypeParameter[] | undefined {
```
Comment (Andarist) [inline src/compiler/checker.ts:10065]: note: this probably should be renamed now, right?

**Comment by Andarist** [comment]
@sandersn I've changed the implementation to include the namespace name now (you can double-check with my [other comment](https://github.com/microsoft/TypeScript/pull/49049#discussion_r876710465) to make sure that this is what is desired here).

**Comment by Andarist** [comment]
Thanks for the review! I can totally dig more into this and compare the codepaths between this case here and the one that you have mentioned.

**Comment by Andarist** [comment]
> For example function f<T>() { function g() { }; return g/**/ } must also need to get an outer type parameter for g, right?

It doesn't appear so because the stringified symbol here is `(local function) g(): void`. We might notice that there is no reference to `T` here.

The outer params are never looked up here because there is only a single symbol in the chain here (`g`) and thus we never get passed [this check](https://github.dev/microsoft/TypeScript/blob/e56a06780141c8ff75f24506572dc35d26d49391/src/compiler/checker.ts#L6081) as this doesn't hold true: `index < (chain.length - 1)`. The `index` is 0 and and the `chain.length` is 1.

So let's check what happens with longer chains:
```ts
function createProps<T>(t: T) {
  const getProps = () => {}
  const createVariants = () => {}
 
  getProps.createVariants = createVariants;
  getProps./**/createVariants
  return getProps;
}
```

This symbol stringifies to `(property) getProps.createVariants: () => void` so it doesn't refer to `T` either. This one passes the mentioned check though but quickly skips over a call to `getTypeParametersOfClassOrInterface` (that calls to `getOuterTypeParametersOfClassOrInterface`) based on [another check](https://github.dev/microsoft/TypeScript/blob/e56a06780141c8ff75f24506572dc35d26d49391/src/compiler/checker.ts#L6084) a few lines below that.

> Your new code looks correct if getOuterTypeParametersOfClassOrInterface is in fact the correct code path.

Based on the analysis above - it seems to me that this is the correct code path. The difference here is that the chain is longer than 1 and that the non-last symbol is instantiated.

File: src/compiler/checker.ts (hunk 2 above)
Comment (sandersn) [reply src/compiler/checker.ts:10065]: My guess is that the correct thing is to instead call `getOuterTypeParameters` at some point, but I'm not sure where.

Comment (Andarist) [reply src/compiler/checker.ts:10065]: > My guess is that the correct thing is to instead call getOuterTypeParameters from lookupTypeParameterNodes

If I understand correctly, you are proposing to exchange the implemented code path:
`lookupTypeParameterNodes -> getTypeParametersOfClassOrInterface -> getOuterTypeParametersOfClassOrInterface -> getOuterTypeParameters`
to
`lookupTypeParameterNodes -> getOuterTypeParameters`.

The difference would be that:
1. we don't call into `getTypeParametersOfClassOrInterface` to eventually reach `getOuterTypeParameters`
2. we wouldn't be calling `getLocalTypeParametersOfClassOrInterfaceOrTypeAlias`

So ye, I definitely agree that, with the current implementation, calling to `getLocalTypeParametersOfClassOrInterfaceOrTypeAlias` is:
- either redundant
- broken too

I've checked if I can (should?) obtain local type params of the same symbol and based on some existing behavior I think I should:
```ts
function create<T>(t: T) {
  return class<T2> {
    someMethod() {}
  };
}

// (method) (Anonymous class)<"outer", "local">.someMethod(): void
new (create("outer" as const))<'local'>()./**/someMethod();
```

We can see here that the local type param of this anonymous class is being printed. This **doesn't** happen with the current state of this PR for those inner expando~ functions.

So given that a local type params should also be obtained here and printed I would have to both obtain inner and outer type params in the `lookupTypeParameterNodes` for this case. Which is already **exactly** what the current code path should be doing (it's just not handling expando functions). 

So I wonder - do we really want to special case this situation here in the `lookupTypeParameterNodes`? Wouldn't it be easier to just introduce the required logic to `getOuterTypeParametersOfClassOrInterface`, `getLocalTypeParametersOfClassOrInterfaceOrTypeAlias` **and** rename them?

A sideways observation is that in the given example with the anonymous class (and in the tests that I've added to this PR) the outer type params are "appended" to the symbol that doesn't reference them at all. We can notice this: `(Anonymous class)<"outer", "local">` but that anonymous class has only a single type param. I somewhat find this confusing but I assume that this is how it was designed to work.

Comment (Andarist) [reply src/compiler/checker.ts:10065]: @sandersn friendly 🏓

File: src/compiler/checker.ts (hunk 1 above)
Comment (sandersn) [reply src/compiler/checker.ts]: I think that omitting `getProps` is OK here, and that the return type of `createProps` is correct. You can't refer to `getProps` on the return value, so it's fine not to name it via generating a namespace.

**Comment by sandersn** [comment]
I looked at the test output again, eg for quickInfoOnFunctionPropertyReturnedFromGenericFunction1.ts. Currently it produces `getProps<{}>.createVariants`; the first draft produced `createVariants`. I *think* the best output is actually `createProps<{}>.createVariants`, since that's the externally visible name.

However, I think I've confused myself pretty thoroughly at this point. @weswigham, you've worked with .d.ts output a lot. What do you think is best in this case?

**Comment by Andarist** [comment]
> I think the best output is actually createProps<{}>.createVariants, since that's the externally visible name.

How is it the externally visible name? I'd like to double-check if I understand your intent here - quite frankly, after all of those months, my memory is also fuzzy as to all the details I was checking when preparing this PR.

When I take a look at quick info that is using the generated `.d.ts` then I can only see this:
```ts
(property) createVariants: () => void
```

**Comment by sandersn** [comment]
Calling `createProps({})` returns an object with a property `createVariants`, so it makes sense to me for quick info to produce `createProps<{}>.createVariants` for quickinfo. But just `createVariants` seems fairly good as well. 
I'm not at all sure which is best for ease of understanding, which seems like the important function of quickinfo.

**Comment by Andarist** [comment]
@sandersn actually, I think that the current version is the correct one - if we consider how other things work.

Consider this example that shared the same code structure but one that is returning a class from a factory:
```ts
function create<T>(t: T) {
  return class InnerName {
    someMethod() {}
  };
}

// (method) InnerName<{}>.someMethod(): void
new (create({}))()./**/someMethod();
```

and this is despite the fact that the generated `.d.ts` doesn't include `InnerName` anyhow:
```ts
declare function create<T>(t: T): {
    new (): {
        someMethod(): void;
    };
};
```

So there are already cases for which you can get different quick info in the source file and in the declaration file.

Since this PR is just about a bug fix and not any kind of cleanup/refactor, I'm inclined to say that this should be merged "as is". Let me know what do you think.

**Review by sandersn — APPROVED** [review]

**Comment by sandersn** [comment]
Thanks for the example. Let's merge this as-is, then.

**Pull request merged**